src/preact/slot.js

```diff
@@ -32,12 +32,20 @@ const EMPTY = {};
 /**
  * @param {!Element} element
  * @param {string} name
- * @param {!Object|undefined} props
- * @return {!PreactDef.VNode}
+ * @param {!Object|undefined} defaultProps
+ * @param {boolean} as
+ * @return {!PreactDef.VNode|!PreactDef.FunctionalComponent}
  */
-export function createSlot(element, name, props) {
+export function createSlot(element, name, defaultProps, as) {
+  /**
+   * @param {!Object|undefined} props
+   * @return {!PreactDef.VNode}
+   */
+  function SlotWithProps(props) {
```

Comment:
This actually has the same issue mentioned in the other thread. Here, `SlotWithProps` function is being recreated every time we render, which means the Component identity changes every time. This will cause the `<slot>` DOM node to be removed, a new one created, and then inserted during every diff.

Now that we're operating on the `PreactBaseElement` level, we can actually cache this per the element instance. Eg:

```js
const cache: WeakMap<Element, {oldDefauls: (!Object|undefined), component: Component}> = new WeakMap();

export function createSlot(element, name, defaultProps, as) {
  element.setAttribute('slot', name);
  if (!as) {
    return <Slot {...(defaultProps || EMPTY)} name={name} />;
  }

  const cached = cache.get(element);
  if (cached && shallowEquals(cached.oldProps, defaultProps)) {
    return cached.component;
  }
  function SlotWithProps(props) {
    return <Slot {...(defaultProps || EMPTY)} name={name} {...props} />;
  }
  cache.set(element, { oldProps: defaultProps, component: SlotWithProps });

  return SlotWithProps;
}

Reply:
Thanks for clarifying! Updated.

Reply:
Looks good. I'd like to see a unit test verifying the `<slot>` DOM node is the same after a rerender, and possibly one asserting that it is new after the `defaultProps` changes.

Reply:
To get the second case, we will have to enable both deep prop def checking in `shouldMutationBeRerendered` and include `subtree: true` in `CHILDREN_MUTATION_INIT` -- wanted to confirm this is OK.

Reply:
I think we can avoid it with:

```js
PBE['props'] = {
  next: {
    selector: '…',
    props: {
      bar: {attr: 'bar'},
    },
  },
  foo: {
    attr: 'foo',
  },
};

// Mutate slot prop, but this won't trigger a rerender
el.querySelector('next').setAttribute('bar', 'other');

// Mutate an observed attr to trigger rerender
el.setAttribute('foo', 'foooey');
```

If it's more complicated than that, I think it's ok with just the unit tests you wrote.

Reply:
Oh nice, that works. I removed the changes to the mutation observer then. Thanks!